codesign/lib/src/file_codesign_visitor.dart

```diff
@@ -0,0 +1,111 @@
+// Copyright 2019 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+import 'package:codesign/codesign.dart';
+import 'package:file/file.dart';
+import 'package:process/process.dart';
+
+/// Interface for classes that interact with files nested inside of [RemoteZip]s.
+abstract class FileVisitor {
+  const FileVisitor();
+
+  Future<void> visitEmbeddedZip(EmbeddedZip file, String parent, String entitlementParentPath);
+  Future<void> visitRemoteZip(RemoteZip file, Directory parent);
+  Future<void> visitBinaryFile(BinaryFile file, String entitlementParentPath);
+}
+
+enum NotaryStatus {
+  pending,
+  failed,
+  succeeded,
+}
+
+/// Codesign and notarize all files within a [RemoteArchive].
+class FileCodesignVisitor extends FileVisitor {
+  FileCodesignVisitor({
+    required this.tempDir,
+    required this.commitHash,
+    required this.processManager,
+    required this.codesignCertName,
+    required this.codesignPrimaryBundleId,
+    required this.codesignUserName,
+    required this.appSpecificPassword,
+    required this.codesignAppstoreId,
+    required this.codesignTeamId,
+    required this.stdio,
+    required this.filepaths,
+    this.production = false,
+  });
+
+  /// Temp [Directory] to download/extract files to.
+  ///
+  /// This file will be deleted if [validateAll] completes successfully.
+  final Directory tempDir;
+
+  final String commitHash;
+  final ProcessManager processManager;
+  final String codesignCertName;
+  final String codesignPrimaryBundleId;
+  final String codesignUserName;
+  final String appSpecificPassword;
+  final String codesignAppstoreId;
+  final String codesignTeamId;
+  final Stdio stdio;
+  final bool production;
+  final List<String> filepaths;
+  // Utility utility = Utility();
+  // TODO(xilaizhang): add back utitlity in later splits
+  Set<String>? fileWithEntitlements;
+  Set<String>? fileWithoutEntitlements;
```

Comment:
I'm guessing these will be used in a later PR.  To avoid checking for these being empty all the time it's probably better to initialize them as empty, unless them being null is meaningful.
```suggestion
  final Set<String> fileWithEntitlements = <String>{};
  final Set<String>? fileWithoutEntitlements = <String>{};
```